### PR TITLE
chore(hybridcloud) Remove role switching and update guard method

### PR DESCRIPTION
**Pull request by markstory**
The role switching logic to catch write operations that are missing outboxes has been causing test flakiness and disrupting development workflows.

These changes finish replacing role switching with query auditing that is done during test teardown. Switching to a teardown based audit means we can't give the precise location of where the bad operation was as the stack frames have been lost. However, we are able to show the offending query and its location in the SQL logs for a test.

### CODE REVIEW

File: src/sentry/db/postgres/roles.py [hunk 1]
```diff
@@ -1,42 +1,8 @@
...
 
-from django.db.transaction import get_connection
-
-from sentry.silo.patches.silo_aware_transaction_patch import determine_using_by_silo_mode
-
-_fencing_counters: MutableMapping[str, int] = defaultdict(int)
-
-
-@contextlib.contextmanager
 def in_test_psql_role_override(role_name: str, using: str | None = None):
```
Comment (markstory) [inline src/sentry/db/postgres/roles.py:4]: This will be around for a bit longer until I can update getsentry.

File: src/sentry/models/counter.py [hunk 2]
```diff
@@ -94,6 +93,8 @@ def increment_project_counter(project, delta=1, using="default"):
 # this must be idempotent because it seems to execute twice
 # (at least during test runs)
 def create_counter_function(app_config, using, **kwargs):
+    from sentry.testutils.silo import unguarded_write
```
Comment (markstory) [inline src/sentry/models/counter.py:96]: Circular import sadness.

File: src/sentry/testutils/asserts.py [hunk 3]
```diff
@@ -1,7 +1,5 @@
 from typing import Optional
 
-from sentry.models import AuditLogEntry, CommitFileChange
```
Comment (markstory) [inline src/sentry/testutils/asserts.py:3]: More circular import sadness.

File: src/sentry/testutils/silo.py [hunk 4]
```diff
@@ -231,129 +234,79 @@ def test_something(self):
...
-        conn.execute(f"GRANT USAGE ON SCHEMA public TO {role};")
-        conn.execute(f"GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO {role};")
-        conn.execute(f"GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public TO {role};")
-
-
-_role_created: bool = False
-_role_privileges_created: MutableMapping[str, bool] = {}
 
+@contextlib.contextmanager
+def unguarded_write(using: str | None = None, *args: Any, **kwargs: Any):
```
Comment (markstory) [inline src/sentry/testutils/silo.py:241]: I'm not :100: sold on this name, and welcome any suggestions to improve it.

File: src/sentry/testutils/silo.py [hunk 5]
```diff
@@ -231,129 +234,79 @@ def test_something(self):
...
+    for app_config in apps.get_app_configs():
+        for model in iter_models(app_config.name):
+            for field in model._meta.fields:
+                if not isinstance(field, HybridCloudForeignKey):
+                    continue
+                fk_model = field.foreign_model
+                if fk_model is None or fk_model in seen_models:
+                    continue
+                seen_models.add(fk_model)
+                _protected_operations.append(protected_table(fk_model._meta.db_table, "delete"))
```
Comment (markstory) [inline src/sentry/testutils/silo.py:293]: I missed adding these guards in the initial pass, but noticed these were missing when removing the role based methods.

File: tests/sentry/models/test_integration.py [hunk 6]
```diff
@@ -10,14 +7,6 @@
 
 @control_silo_test(stable=True)
 class IntegrationTest(TestCase):
-    def test_cannot_delete_with_queryset(self):
```
Comment (markstory) [inline tests/sentry/models/test_integration.py:13]: These tests had to be removed as they now throw during teardown which we can't catch.

File: src/sentry/db/postgres/roles.py [hunk 7]
```diff
@@ -1,42 +1,8 @@
...
-    if "pytest" not in sys.argv[0] or os.environ.get("DB", "postgres") != "postgres":
-        yield
-        return
-
-    using = determine_using_by_silo_mode(using)
-
-    # TODO(mark) Move this closer to other silo code.
-    _fencing_counters[using] += 1
+    # Deprecated shim for getsentry compatibility
+    from sentry.testutils.silo import unguarded_write
```
Comment (corps) [inline src/sentry/db/postgres/roles.py:6]: does this need the `using = determine_using_by_silo_mode` logic for split db still?

Comment (corps) [reply src/sentry/db/postgres/roles.py:6]: Nevermind, it is captured inside the call.

File: src/sentry/testutils/silo.py (hunk 5 above)
Comment (corps) [reply src/sentry/testutils/silo.py:293]: There should also be a test, somewhere, I wrote that fails if a raw deletion succeeds without throwing an exception.

**Review by corps — APPROVED** [review]

File: tests/sentry/models/test_integration.py (hunk 6 above)
Comment (corps) [reply tests/sentry/models/test_integration.py:13]: Is there some alternative that ensures atleast that this logic does exist?

**Review by dashed — APPROVED** [review]

**Review by GabeVillalobos — APPROVED** [review]

**Pull request merged**